### ISSUE TITLE
current devel branch won't compile on mac osx

### ISSUE THREAD

**Issue by ppxasjsm** [michellab/Sire#139]
I did a clean checkout and install of sire on my mac, but it will fail, when compiling on my Mac. It compiles fine on Ubuntu 14.04, but then has these import errors reported in issue #136.  
Here is part of the error:

```
<-1> >, boost::python::objects::register_base_of<SireCAS::AlwaysFalse> >(boost::mpl::v_iter<boost::python::bases<SireCAS::Condition, SireCAS::ExBase, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_>, 1l>*, boost::mpl::v_iter<boost::python::bases<SireCAS::Condition, SireCAS::ExBase, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_>, 2l>*, boost::add_pointer<mpl_::arg<-1> >*, boost::python::objects::register_base_of<SireCAS::AlwaysFalse>) in AlwaysFalse.pypp.cpp.o
      register_AlwaysTrue_class() in AlwaysTrue.pypp.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CAS/_CAS.2016.3.1.so] Error 1
make[1]: *** [CAS/CMakeFiles/CAS.dir/all] Error 2
make: *** [all] Error 2
SOMETHING WENT WRONG WHEN COMPILING WRAPPER!
```

I am trying to make some changes for the new release, but devel does not seem to be safe to branch off from at the moment...

**Comment by chryswoods** [comment]
I'll try a clean compile and install now. Nothing has changed on devel since the 2016.3.1 release. I've been working in a the new_parser branch and am being strict with myself now not to do anything other than use pull-requests into devel.

This is a very strange error. Which OS X are you using, and which compiler?

**Comment by ppxasjsm** [comment]
OSX is: 10.10.5

```Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.0 (clang-700.0.72)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```

**Comment by ppxasjsm** [comment]
I haven't changed anything since the last clean build using 2016.2 I think.

**Comment by chryswoods** [comment]
Both my test Macs are 10.12.3 (LLVM 8.0.0). My older test mac had its battery pop, so I lost that. The only major change in linking etc from 2016.2 to 2016.3 (indeed 2016.3 to 2016.3.1) is that I have had to switch on Mac backwards-compatibility checking. This was achieved by setting the environment variable "export MACOSX_DEPLOYMENT_TARGET="10.8"" in compile_sire.sh. This ensured that all libraries were compiled using symbols that were available for 10.8. 

(my clean compile is still building Qt...)

**Comment by chryswoods** [comment]
I now get the same build error on my MacBook. Very strange, as I don't see this with the continued build in new_parser... The errors look like the boost::python library has either not built properly or has moved. I suspect it may be caused by a change in miniconda.sh. I'll know more after investigating...

**Comment by chryswoods** [comment]
Currently looking like it may be a fake gcc / clang issue (https://github.com/GalSim-developers/GalSim/issues/483). I'm now trying to recompile using;

CC=clang CXX=clang++ ./compile_sire.sh

and will let you know how I get on. Also, as an aside, the new miniconda.sh is Python 3.6, while the working miniconda.sh on my development mac is 3.5.2. Don't know if this means anything

**Comment by chryswoods** [comment]
No, that didn't fix things... I will continue to look at this tomorrow. I think I will try with an older miniconda to see if that was the problem...

**Comment by ppxasjsm** [comment]
thank you for looking into this.

**Comment by chryswoods** [comment]
No, using an older Python version didn't work... This is a very peculiar problem, as compiling new_parser, which is forked from the latest devel, is fine...

**Comment by chryswoods** [comment]
Now trying to compile from the master branch (exact equal to 2016.3.1) to validate that it can compile... (it definitely used to, as this was part of the testing pre-christmas)

**Comment by chryswoods** [comment]
No - even master is not compiling any more. This is getting frustrating...

**Comment by ppxasjsm** [comment]
I had also tried a 2016.0 release tag and tried to compile it. Issue #137 addresses this, I have no idea why none of them work. For sure the 2016.0 reliably compiled both on my mac and linux machines.

**Comment by chryswoods** [comment]
Curious... the new_parser branch compiles well. I suspect there was a change in OS X that caused the link errors and I fixed this in new_parser. I will look for the differences and will back port the fix to devel.

**Comment by chryswoods** [comment]
Nothing obvious pops out from the diff between branches (https://github.com/michellab/Sire/compare/new_parser). I am now compiling devel as a separate instance and will manually compare against new_parser. This is very strange.

**Comment by chryswoods** [comment]
Ok - seen that libboost_python.dylib in new_parser is linked against libc++.1.dylib, while it is linked against libstdc++.6.dylib. This is the cause of the problem, as the wrong c++ library has been picked up. However, how new_parser is getting the right library and devel is not is an open question... I can't find any difference in the cmake files, compile_sire.sh, build_sire.py etc. that fixes this.

**Comment by chryswoods** [comment]
Ok, I think I have fixed it. I have pushed the change to devel and it is compiling on travis and I've test complied on my MacBook (and am now test compiling from scratch to double-check). The problem was that - somehow - all branches of Sire except for new_parser followed correct behaviour of setting MAC_OSX_DEPLOYMENT_TARGET of 10.8 and setting up the build path. The install_boost_python.cmake script did not set the correct C++ library to use as it used the incorrect variable "SIRE_CXX11_FLAGS", which didn't exist anywhere else. This caused boost::python to default link against libstdc++ rather than libc++, hence the missing symbol link errors. Why this suddenly started happening now, and why it didn't affect new_parser, I do not know. (There is no difference that I can see, except that new_parser manages to ignore MAC_OSX_DEPLOYMENT_TARGET when building boost::python, meaning that it automatically selects libc++ and so works). The fix is that I have added the C++ library to the SIRE_CXX_NG_FLAGS, and made sure that boost::python uses these in its cmake file. Very random!
